src/MessagePack/LZ4/LZ4MessagePackSerializer.JSON.cs

```diff
@@ -157,16 +157,48 @@ static int ToJsonCore(byte[] bytes, int offset, StringBuilder builder)
                         return totalReadSize;
                     }
                 case MessagePackType.Extension:
-                    var ext = MessagePackBinary.ReadExtensionFormat(bytes, offset, out readSize);
-                    if (ext.TypeCode == ReservedMessagePackExtensionTypeCode.DateTime)
+                    var extHeader = MessagePackBinary.ReadExtensionFormatHeader(bytes, offset, out readSize);
+                    if (extHeader.TypeCode == ReservedMessagePackExtensionTypeCode.DateTime)
                     {
                         var dt = MessagePackBinary.ReadDateTime(bytes, offset, out readSize);
                         builder.Append("\"");
                         builder.Append(dt.ToString("o", CultureInfo.InvariantCulture));
                         builder.Append("\"");
                     }
+                    else if (extHeader.TypeCode == ReservedMessagePackExtensionTypeCode.DynamicObjectWithTypeName)
+                    {
+                        int startOffset = offset;
+                        // prepare type name token
+                        offset += 6;
+                        var typeNameToken = new StringBuilder();
+                        var typeNameReadSize = ToJsonCore(bytes, offset, typeNameToken);
+                        offset += typeNameReadSize;
+                        int startBuilderLength = builder.Length;
+                        if (extHeader.Length > typeNameReadSize)
+                        {
+                            // object map or array
+                            var typeInside = MessagePackBinary.GetMessagePackType(bytes, offset);
```

Comment:
It was not intended for map or array directly, but for injecting $type into serialized object fields, as object fields are serialized as map or array, depending on parameter - KeyAsString or KeyAsInt.